pycloudlib/azure/cloud.py

```diff
@@ -0,0 +1,644 @@
+# This file is part of pycloudlib. See LICENSE file for license information.
+"""Azure Cloud type."""
+import base64
+
+from azure.mgmt.resource import ResourceManagementClient
+from azure.mgmt.network import NetworkManagementClient
+from azure.mgmt.compute import ComputeManagementClient
+
+import pycloudlib.azure.util as util
+
+from pycloudlib.cloud import BaseCloud
+from pycloudlib.azure.instance import AzureInstance
+from pycloudlib.key import KeyPair
+from pycloudlib.util import create_tag
+
+
+class Azure(BaseCloud):
+    """Azure Cloud Class."""
+
+    _type = 'azure'
+
+    UBUNTU_RELEASE = {
+        "trusty": "Canonical:UbuntuServer:14.04.0-LTS",
+        "xenial": "Canonical:UbuntuServer:16.04-DAILY-LTS",
+        "bionic": "Canonical:UbuntuServer:18.04-DAILY-LTS",
+        "focal": "Canonical:0001-com-ubuntu-server-focal-daily:20_04-daily-lts"
+    }
+
+    def __init__(self, tag, client_id=None, client_secret=None,
+                 subscription_id=None, tenant_id=None):
+        """Initialize the connection to Azure.
+
+        Azure will try to read user credentials form the /home/$USER/.azure
+        folder. However, we can overwrite those credentials with the provided
+        id parameters.
+
+        Args:
+            tag: string used to name and tag resources with
+            client_id: user's client id
+            client_secret: user's client secret access key
+            subscription_id: user's subscription id key
+            tenant_id: user's tenant id key
+        """
+        super().__init__(tag)
+        self._log.debug('logging into Azure')
+        self.location = "centralus"
+        self.username = "ubuntu"
+
+        self.registered_instances = {}
```

Comment:
I don't think this approach will work for other pycloudlib.Azure users as this only gets set as you launch instances on this particular run of pycloudlib.Azure. We also want to be able to query existing instances in the cloud (not launched by this particular instance of pycloudlib). I think we should use the compute api for this compute_client.virtual_machines.list(resource_group_name='blah')

Reply:
Yes, you are right. I was thinking about the resources only allocated by the instance, but we should think globally here. I will update the code